README.md
peakAnalysis
============

R package for peak analysis

Author: Brian Jones Laboratory. The University of Sydney.   
Contributors: Mario Fajardo, Juan Pablo Matte and Jose Padarian

Contains functions for peak analysis




NEWS
peak analysis 0.1
--------------------------------------------------------------
* Added 2  functions

Test_1.R
####Install Bioconductor libraries and required packages####
source("http://bioconductor.org/biocLite.R")
## Use biocLite()

biocLite(c("GenomicFeatures", "AnnotationDbi","flowCore",'flowViz'))

#####Read FACS data####

require(flowCore)

# in some cases we use pattern = 'arf', to select some of the files
# Col0_files<-dir('C:/Users/lab/Documents/JUAN PABLO/FACS_Data/External JuanM20141216/',pattern = 'Col0',full.names = T)

FACSC20141216_files <- dir('C:/Users/lab/Documents/JUAN PABLO/FACS_Data/External JuanM20141216/',full.names = T)



#read the files
FACSC20141216 <- lapply(FACSC20141216_files,function(x) read.FCS(x, transformation=FALSE))


#check properties of each object#
summary(FACSC20141216[[1]])
FACSC20141216[[1]]@exprs[1:10,1:10]

# in the library of FACS says that we need to check if there is a parameter > 0, but we didn't find any
#check if this is higher than 0 #
#this is a test of one of them
keyword(object = FACSC20141216[[1]],paste0('$P1E'))


#alter names for mor "R friendly names"#
#note ...the function alter the names first and after that looks for a column pattern#
FACSC20141216<- lapply(FACSC20141216_files,function(x) read.FCS(x, transformation=FALSE,alter.names = T))
FACSC20141216[[1]]@exprs[1:5,1:5]

#Visualization#

require('flowViz')
#DON'T DO THIS...TO MANY GRAPHICS#
#plot(FACSC20141216[[1]])

FACSC20141216[[1]]

par()
plot(FACSC20141216[[1]],c("PI.H", "DAPI.H"))
#Would be better to use log scale in both axes


require(ggplot2)

#example of plot# the points in the middle are the nuclei
ggplot(as.data.frame(FACSC20141216[[1]]@exprs),aes(PI.H,DAPI.H))+
  geom_point()+
  scale_x_log10()+
  scale_y_log10()

#example of histogram# but if we plot all, we got the peaks and noise
ggplot(as.data.frame(FACSC20141216[[1]]@exprs),aes(PI.H))+
  geom_density()+
  scale_x_log10()
data <-as.data.frame(FACSC20141216[[1]]@exprs[,c('PI.H','DAPI.H')])

# Using manipulate we treshole the data
require(manipulate)

manipulate(
  ggplot(data[data$PI.H/data$DAPI.H>=down & round(data$PI.H/data$DAPI.H,3)<=up,],aes(PI.H,DAPI.H))+
    geom_point(fill='red',alpha=.4)+
    scale_y_log10(limits=c(100, 80000))+
    scale_x_log10(limits=c(100, 80000)),  
  down=slider(min=.7,max=0.89,step=.01),
  up=slider(min = 0.9,max = 1.5,step = .01))

# And then the same with the histogram, the peaks are much more clear

manipulate(
   ggplot(data[data$PI.H/data$DAPI.H>=down & round(data$PI.H/data$DAPI.H,3)<=up,],aes(PI.H))+
    geom_density(fill='red',alpha=.4)+
    scale_x_log10(limits=c(100, 80000)),  
  down=slider(min=.7,max=0.89,step=.01),
  up=slider(min = 0.9,max = 1.5,step = .01))

# Then the next step is to find a way to optimize this threshold checking all the possible combination
# and get the local max, but using the 2D histogram information.

#Next file is Pea_analysis

peaks.R
library(raster)
require(flowCore)
require(maps)

FACSC20141216_files <- dir('.','.fcs',full.names = T)

findPeaks <- function(file,silent=F) {
  data <- read.FCS(file, transformation=FALSE,alter.names = T)
  to_plot <- data.frame(x=data@exprs[,'PI.H'],y=data@exprs[,'DAPI.H'])
  to_plot <- data_ <- log(to_plot) 
  
  coordinates(data_) <- ~ x + y
  
  template <- raster(res=0.1,ext=extent(data_),crs=NULL)
  test <- rasterize(data_,template,data_@coords[,1],fun='count')
  
  test_ <- na.exclude(as.data.frame(cbind(xyFromCell(test,cell = 1:ncell(test)),test[])))
  lines <- test_[test_[,3] > quantile(test_[,3],0.995,na.rm = T),1:2]
    
  simplify <- function(lines) {
    aaa <- lapply(1:nrow(lines), function(x) {
      test <- fields::rdist(lines[x,],lines[-x,]) <= 0.2
      sort(c(rownames(lines[x,]),rownames(lines[-x,])[test]))
    })
    
    if(nrow(lines[sapply(aaa,length) != 1,]) == 0) return(lines)
    
    bbb <- lines[sapply(aaa,length) == 1,]
    
    aaaa <- sapply(max(sapply(aaa,length)):2,function(x){
      step <- aaa[sapply(aaa,length) == x]
      if(length(step) == 0) return(NULL)
      apply(unique(do.call(rbind,step)),1,function(y){
        colMeans(lines[y,])
      })
    })
    if(class(aaaa) == 'list') {
      aaaaa <- t(do.call(cbind,aaaa))
    } else {
      aaaaa <- t(aaaa)
    }
    
    colnames(aaaaa) <- colnames(bbb) <- c('x','y')
    
    simplify(rbind(aaaaa,bbb))
  }
  
  cc <- simplify(lines)
  
  buffer <- extract(test,cc,buffer=0.25,fun=sum)
  values <- extract(test,cc,fun=mean)
  
  
      plot(test,xlab='log(PI-H)',ylab='log(DAPI-H)')
      sapply(cc[,1],function(x) abline(v=x,lty=2))
      sapply(cc[,2],function(x) abline(h=x,lty=2))
      #apply(cc,1,function(x) points(x=x[1],y=x[2],col='blue',cex=2))
      symbols(cc,circles=rep(0.25,nrow(cc)),inches=F,add=T)
  #cat(c("Peaks' coordinates:",apply(as.matrix(cc),1,function(x) paste0(round(x,2),collapse = ', '))),sep = '\n')
  response <- cbind(cc,peak.value=values,buffer)
  colnames(response) <- c('log(PI-H)','log(DAPI-H)','peak','buffer_0.25u')
  if(silent) response else {
    cat(paste0('***** ',basename(file),' *****\n'))
    print(response) 
  }
}


findPeaks(FACSC20141216_files[[1]])



experiment.R
require(flowCore)
require(MASS)
require(ggplot2)
require(reshape2)
require(Rwave)
require(EBImage)

FACSC20141216_files <- dir('FACS_02FEB2015/.','.fcs',full.names = T)

findPeaks1 <- function(file,tolerance=.995,peaks=5, resol=250,plot=F) {

  data <- read.FCS(file, transformation=FALSE,alter.names = T)
  mask<-data@exprs[,'PI.H']/data@exprs[,'DAPI.H']>=0.8&data@exprs[,'PI.H']/data@exprs[,'DAPI.H']<=1.2
  to_plot <- data.frame(x=data@exprs[,'PI.H'],y=data@exprs[,'DAPI.H'])[mask,]
#   plot(log(to_plot$x),log(to_plot$y))
  kern<-kde2d(log(to_plot$x),log(to_plot$y),n = resol)

  mask<-kern$z*as.numeric(kern$z>quantile(kern$z,tolerance))
  
  to_plot_kern<-data.frame(x=rep(kern$x,resol),y=rep(kern$y,each=resol),z=melt(kern$z)[,3])
  
#this doesn't make sense to me, it should be colMeans...but is not ... i need some sleep#
#REMINDER: always read the help file!... 
#kern2d creates a matrix were x are rows....now peace reigns again#
  loc_max<-function(DATA) which(diff(sign(diff(DATA)))==-2)+1
  loc_max_posit_x<-loc_max(rowSums(kern$z*mask))
  local_maximas_x<-kern$x[loc_max_posit_x]
  if(length(na.exclude(local_maximas_x))<peaks) {
    warning(immediate. = T,call.=FALSE,paste('only',length(na.exclude(local_maximas_x)),' peaks were higher that the tolerance value'))
    local_maximas_x<-local_maximas_x[complete.cases(local_maximas_x)]
    peaks<-length(local_maximas_x)
  }
  loc_max_posit_y<-sapply(loc_max_posit_x[order(local_maximas_x)[1:peaks]],function(x) tmp<-which.max(kern$z[x,]))
  
  local_maximas_y<-sapply(loc_max_posit_x[order(local_maximas_x)[1:peaks]],function(x) {
    tmp<-which.max(kern$z[x,])
    kern$y[tmp]
  })
  local_maximas_x<-kern$x[loc_max_posit_x][1:peaks]

# ggplot(to_plot,aes(log(x),log(y)))+
#   geom_point()+
#   coord_equal()+
#   labs(x='log(PI-H)',y='log(DAPI-H)')
if(plot){
    print(ggplot(to_plot_kern,aes(x,y))+
    geom_tile(aes(fill=z))+
    coord_equal()+
    labs(x='log(PI-H)',y='log(DAPI-H)')+
    geom_vline(xintercept=local_maximas_x,linetype=2,colour=local_maximas_x)+
    geom_hline(yintercept=local_maximas_y,linetype=2,colour=local_maximas_x))
}
# pix_size_x <- abs(kern$x[1]- kern$x[2])
# pix_size_y <- abs(kern$y[1]- kern$y[2])
# 
# 
# buffer <- mapply(FUN = function(x,y) kern$z[x-x*(buffer_size/pix_size_x):x-buffer_size/pix_size_x,y],loc_max_posit_x,loc_max_posit_y)
results <- round(data.frame(x_value=local_maximas_x,y_value=local_maximas_y),3)
colnames(results) <- c('log(PI-H)','log(DAPI-H)')

cat(paste0('Peaks for sample: \n##', basename(file), '##'),'\n')
print(results)
cat('#################################')
}  

findPeaks1(file = FACSC20141216_files[[4]],tolerance = .98,resol = 100,peaks = 5,plot = T)



R/findPeaks1.R
#' findPeaks1
#' 
#' This function looks for peaks in dna files 
#' @param file a dna file
#' @param tolerance limit for the peak detection (need to be more specific...)
#' @param peaks number of desired peaks to find
#' @param resol resolution of the kernel, default to 250 
#' @param plot default TRUE
#' @importFrom flowCore read.FCS
#' @importFrom MASS kde2
#' @import ggplot2
#' @importFrom reshape2 melt 
#' @import ggplot2
#' @export findPeaks1




findPeaks1 <- function(file,tolerance=.995,peaks=5, resol=250,plot=F) {

  data <- read.FCS(file, transformation=FALSE,alter.names = T)
  mask<-data@exprs[,'PI.H']/data@exprs[,'DAPI.H']>=0.8&data@exprs[,'PI.H']/data@exprs[,'DAPI.H']<=1.2
  to_plot <- data.frame(x=data@exprs[,'PI.H'],y=data@exprs[,'DAPI.H'])[mask,]
#   plot(log(to_plot$x),log(to_plot$y))
  kern<-kde2d(log(to_plot$x),log(to_plot$y),n = resol)

  mask<-kern$z*as.numeric(kern$z>quantile(kern$z,tolerance))
  
  to_plot_kern<-data.frame(x=rep(kern$x,resol),y=rep(kern$y,each=resol),z=melt(kern$z)[,3])
  
#this doesn't make sense to me, it should be colMeans...but is not ... i need some sleep#
#REMINDER: always read the help file!... 
#kern2d creates a matrix were x are rows....now peace reigns again#
  loc_max<-function(DATA) which(diff(sign(diff(DATA)))==-2)+1
  loc_max_posit_x<-loc_max(rowSums(kern$z*mask))
  local_maximas_x<-kern$x[loc_max_posit_x]
  if(length(na.exclude(local_maximas_x))<peaks) {
    warning(immediate. = T,call.=FALSE,paste('only',length(na.exclude(local_maximas_x)),' peaks were higher that the tolerance value'))
    local_maximas_x<-local_maximas_x[complete.cases(local_maximas_x)]
    peaks<-length(local_maximas_x)
  }
  loc_max_posit_y<-sapply(loc_max_posit_x[order(local_maximas_x)[1:peaks]],function(x) tmp<-which.max(kern$z[x,]))
  
  local_maximas_y<-sapply(loc_max_posit_x[order(local_maximas_x)[1:peaks]],function(x) {
    tmp<-which.max(kern$z[x,])
    kern$y[tmp]
  })
  local_maximas_x<-kern$x[loc_max_posit_x][1:peaks]

# ggplot(to_plot,aes(log(x),log(y)))+
#   geom_point()+
#   coord_equal()+
#   labs(x='log(PI-H)',y='log(DAPI-H)')
if(plot){
    print(ggplot(to_plot_kern,aes(x,y))+
    geom_tile(aes(fill=z))+
    coord_equal()+
    labs(x='log(PI-H)',y='log(DAPI-H)')+
    geom_vline(xintercept=local_maximas_x,linetype=2,colour=local_maximas_x)+
    geom_hline(yintercept=local_maximas_y,linetype=2,colour=local_maximas_x))
}
# pix_size_x <- abs(kern$x[1]- kern$x[2])
# pix_size_y <- abs(kern$y[1]- kern$y[2])
# 
# 
# buffer <- mapply(FUN = function(x,y) kern$z[x-x*(buffer_size/pix_size_x):x-buffer_size/pix_size_x,y],loc_max_posit_x,loc_max_posit_y)
results <- round(data.frame(x_value=local_maximas_x,y_value=local_maximas_y),3)
colnames(results) <- c('log(PI-H)','log(DAPI-H)')

cat(paste0('Peaks for sample: \n##', basename(file), '##'),'\n')
print(results)
cat('#################################')
}  




Peak_analysis.R
require(flowCore)

FACSC20141216_files <- dir(pattern = 'fcs',full.names = T)

FACSC20141216 <- lapply(FACSC20141216_files,function(x) read.FCS(x, transformation=FALSE,alter.names = T))

localMax<-function(tt) which(diff(sign(diff(tt)))==-2)+1
require(plyr)

results_arf<-lapply(1:length(FACSC20141216_files),function(j){
  
data <-as.data.frame(FACSC20141216[[j]]@exprs[,c('PI.H','DAPI.H')])

up<-seq(0.6,1.05,0.01)
down<-seq(0.5,1,0.01)
combinations<-expand.grid(up,down)
combinations<-combinations[combinations[,1]>combinations[,2],]
combinations<-combinations[combinations[,1]-combinations[,2]>=0.1,]


results<-adply(combinations,1,function(i){
  selection<-data[data$PI.H/data$DAPI.H>=as.numeric(i[2]) & round(data$PI.H/data$DAPI.H,3)<=as.numeric(i[1]),]
  localm <- localMax(density(selection$PI.H,kernel='gaussian')$y)
  local_maximas<-density(selection$PI.H)$x[localm][1:5]
})

#So here we test all the combination of the threshold and then we use the peaks that repeat more,
# but if you have other idea, use it.

final<-sapply(3:7,function(k) table(round(results[,k],0))[which.max(table(round(results[,k],0)))])

})

#MArio was tire and wrote this long chorizo

FACSC20141216_RESULTS<-data.frame(file=dir(pattern = 'fcs'),do.call(rbind,lapply(results_arf,function(x) as.numeric(names(x)))))

colnames(FACSC20141216_RESULTS) <-c('File',1:5)
FACSC20141216_RESULTS

#This is a good approach, but we force to find 5 peaks and probably will be better to find the real ono.
#Also I need the number of how many points we have on each peak, so I can know the % of the total.


R/findPeaks.R
#' findPeaks
#' 
#' This function looks for peaks in dna files
#'  
#' @param file a dna file
#' @param silent should plot be skipped default F 
#' @importFrom flowCore read.FCS
#' @importFrom raster raster rasterize extract extent xyFromCell ncell
#' @import sp
#' @importFrom fields rdist
#' @export findPeaks


findPeaks <- function(file,silent=F) {
  data <- read.FCS(file, transformation=FALSE,alter.names = T)
  to_plot <- data.frame(x=data@exprs[,'PI.H'],y=data@exprs[,'DAPI.H'])
  to_plot <- data_ <- log(to_plot) 
  
  coordinates(data_) <- ~ x + y
  
  template <- raster(res=0.1,ext=extent(data_),crs=NULL)
  test <- rasterize(data_,template,data_@coords[,1],fun='count')
  
  test_ <- na.exclude(as.data.frame(cbind(xyFromCell(test,cell = 1:ncell(test)),test[])))
  lines <- test_[test_[,3] > quantile(test_[,3],0.995,na.rm = T),1:2]
  
  simplify <- function(lines) {
    aaa <- lapply(1:nrow(lines), function(x) {
      test <- fields::rdist(lines[x,],lines[-x,]) <= 0.2
      sort(c(rownames(lines[x,]),rownames(lines[-x,])[test]))
    })
    
    if(nrow(lines[sapply(aaa,length) != 1,]) == 0) return(lines)
    
    bbb <- lines[sapply(aaa,length) == 1,]
    
    aaaa <- sapply(max(sapply(aaa,length)):2,function(x){
      step <- aaa[sapply(aaa,length) == x]
      if(length(step) == 0) return(NULL)
      apply(unique(do.call(rbind,step)),1,function(y){
        colMeans(lines[y,])
      })
    })
    if(class(aaaa) == 'list') {
      aaaaa <- t(do.call(cbind,aaaa))
    } else {
      aaaaa <- t(aaaa)
    }
    
    colnames(aaaaa) <- colnames(bbb) <- c('x','y')
    
    simplify(rbind(aaaaa,bbb))
  }
  
  cc <- simplify(lines)
  
  buffer <- extract(test,cc,buffer=0.25,fun=sum)
  values <- extract(test,cc,fun=mean)
  
  
  plot(test,xlab='log(PI-H)',ylab='log(DAPI-H)')
  sapply(cc[,1],function(x) abline(v=x,lty=2))
  sapply(cc[,2],function(x) abline(h=x,lty=2))
  #apply(cc,1,function(x) points(x=x[1],y=x[2],col='blue',cex=2))
  symbols(cc,circles=rep(0.25,nrow(cc)),inches=F,add=T)
  #cat(c("Peaks' coordinates:",apply(as.matrix(cc),1,function(x) paste0(round(x,2),collapse = ', '))),sep = '\n')
  response <- cbind(cc,peak.value=values,buffer)
  colnames(response) <- c('log(PI-H)','log(DAPI-H)','peak','buffer_0.25u')
  if(silent) response else {
    cat(paste0('***** ',basename(file),' *****\n'))
    print(response) 
  }
}



